File: android/src/main/kotlin/com/klarna/inapp/sdk/flutter_klarna_inapp_sdk/hybrid/KlarnaHybridSDKHandler.kt
package com.klarna.inapp.sdk.flutter_klarna_inapp_sdk.hybrid

import com.klarna.inapp.sdk.flutter_klarna_inapp_sdk.ResultError
import com.klarna.inapp.sdk.flutter_klarna_inapp_sdk.core.handler.BaseMethodHandler
import com.klarna.mobile.sdk.api.KlarnaEventListener
import com.klarna.mobile.sdk.api.hybrid.KlarnaHybridSDK
import io.flutter.plugin.common.MethodChannel

internal object KlarnaHybridSDKHandler : BaseMethodHandler<KlarnaHybridSDKMethod>(KlarnaHybridSDKMethod.Parser) {

    internal var hybridSDK: KlarnaHybridSDK? = null
    private val hybridSDKCallback = KlarnaHybridSDKCallback()
    private val hybridSDKEventListener: KlarnaEventListener = KlarnaHybridSDKEventListener()

    const val NOT_INITIALIZED = "KlarnaHybridSDK is not initialized"

    internal fun notInitialized(result: MethodChannel.Result?) {
        result?.error(
                ResultError.HYBRID_SDK_ERROR.errorCode,
                NOT_INITIALIZED,
                "Call 'KlarnaHybridSDK.initialize' before this."
        )
    }

    override fun onMethod(method: KlarnaHybridSDKMethod, result: MethodChannel.Result) {
        when (method) {
            is KlarnaHybridSDKMethod.Initialize -> initialize(method, result)
            is KlarnaHybridSDKMethod.RegisterEventListener -> registerEventListener(method, result)
        }
    }

    private fun initialized() = hybridSDK != null

    private fun initialize(method: KlarnaHybridSDKMethod.Initialize, result: MethodChannel.Result) {
        if (hybridSDK == null) {
            hybridSDK = KlarnaHybridSDK(method.returnUrl, hybridSDKCallback)
            result.success(null)
            return
        }
        result.error(ResultError.HYBRID_SDK_ERROR.errorCode, "Already initialized.", null)
    }

    private fun registerEventListener(method: KlarnaHybridSDKMethod.RegisterEventListener, result: MethodChannel.Result) {
        if (!initialized()) {
            notInitialized(result)
            return
        }
        hybridSDK?.registerEventListener(hybridSDKEventListener)
        result.success(null)
    }
}

File: ios/Classes/Hybrid/KlarnaHybridSDKHandler.swift
import Flutter
import KlarnaMobileSDK

class KlarnaHybridSDKHandler: BaseMethodHandler<KlarnaHybridSDKMethod> {
    
    static var hybridSDK: KlarnaHybridSDK? = nil
    static var hybridSDKEventListener: KlarnaHybridEventListener = PluginKlarnaHybridEventListener()
    
    static let NOT_INITIAILIZED = "KlarnaHybridSDK is not initialized"
    
    static func notInitialized(result: FlutterResult?) {
        result?(FlutterError.init(code: ResultError.hybridSdkError.rawValue, message: NOT_INITIAILIZED, details: "Call 'KlarnaHybridSDK.initialize' before using this"))
    }
    
    init() {
        super.init(parser: KlarnaHybridSDKMethods.Parser())
    }
    
    override func onMethod(method: KlarnaHybridSDKMethod, result: @escaping FlutterResult) {
        switch method {
        case is KlarnaHybridSDKMethods.Initialize:
            initialize(method: method as! KlarnaHybridSDKMethods.Initialize, result: result)
        case is KlarnaHybridSDKMethods.RegisterEventListener:
            registerEventListener(method: method as! KlarnaHybridSDKMethods.RegisterEventListener, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func initialize(method: KlarnaHybridSDKMethods.Initialize, result: @escaping FlutterResult) {
        if KlarnaHybridSDKHandler.hybridSDK == nil {
            if let returnUrl = URL.init(string: method.returnUrl) {
                KlarnaHybridSDKHandler.hybridSDK = KlarnaHybridSDK.init(returnUrl: returnUrl, eventListener: KlarnaHybridSDKHandler.hybridSDKEventListener)
                result(nil)
            } else {
                result(FlutterError.init(code: ResultError.hybridSdkError.rawValue, message: "Invalid ReturnURL.", details: nil))
            }
        } else {
            result(FlutterError.init(code: ResultError.hybridSdkError.rawValue, message: "Already initialized.", details: nil))
        }
    }

    private func registerEventListener(method: KlarnaHybridSDKMethods.RegisterEventListener, result: @escaping FlutterResult) {
        guard KlarnaHybridSDKHandler.hybridSDK != nil else {
            result(FlutterError.init(code: ResultError.hybridSdkError.rawValue, message: "Invalid callback for event listener.", details: nil))
            return
        }

        KlarnaHybridSDKHandler.hybridSDK?.registerEventListener(withCallback: { response in
            KlarnaHybridSDKEventHandler.instance.sendValue(value: response.bodyString)
        })

        result(nil)
    }
}
